[ LOPS-1538 ] Cache the terminus directory and only add token if not already logged in

stovak opened the pull request.

stovak committed:
Cache the terminus directory and only add token if not already logged in

G-Rath commented:
fwiw while I've not dig into how Terminus sessions actually work, I expect this will break for repositories that use multiple machine tokens because it isn't included in your cache key.

I'm not sure if it's safe to include it in there either, and unfortunately GHA only supports `hashFiles` so you'd need to write it to a file first to determine its hash

stovak commented:
@G-Rath the problem that we're having is people running a bunch of builds and running afoul of our internal API resource limits. How common is using multiple machine tokens for terminus in a CI environment...??? I can't think of a use-case.

Also, yes, you are correct. It will break that.

G-Rath commented:
With Pantheons current permission model I don't think there is a strong usecase, but I do live in hope it'll be improved at some point to allow us security folk to better lockdown our CIs 😉

But I know that won't be an overnight thing so I mainly mention this for completeness - at this point because of how niche the usecase is it's probably not even worth having an input to allow disabling the cache...

Something else to consider though is the lifetime of the session - am I right in thinking `auth:login` first checks if a valid session exists already, and only actually hits the API if its expired?

greg-1-anderson commented on action.yml:
```diff
@@ -34,11 +34,26 @@ runs:
+          terminus auth:login
+        else
+          terminus auth:login --machine-token="${{ inputs.pantheon-machine-token }}"
+        fi
```
I tried to make this suggestion in Slack, but folks seemed to think I didn't understand the caching component. 😄 

I think that if you rewrote lines 55-59 with `terminus auth:login || terminus auth:login --machine-token="${{ inputs.pantheon-machine-token }}"`, then I think it would work if the cached session does not exist, or if the cached session exists but has expired, or if the cached session exists and is still valid.

The use case for multiple machine tokens is: you revoke the machine token you are using in CI because it was compromised, and you add a new (single) machine token. I think the implementation in the current version of this PR fails when the cached session expires, because there's no good way to get the "--machine-token" code path to run as long as the cached session exists.

greg-1-anderson replied:
Note that I have not tried my suggestion; it only works if `terminus auth:login` returns a status code 1 when the cached session does not exist or has expired. Hopefully it does that.

G-Rath replied:
I like all of this - I can't believe that use-case for multiple machine tokens didn't occur to me, because we have a slightly alternative version: since right now machine tokens are per-user rather than per-site, and there's a high overhead in maintaining a dedicated "machine user" for each site, we've got a single token as an organisation secret so that we can rotate it as often as we want (yearly, monthly, weekly or even daily)

stovak committed:
Implement as single line rather than multiline bash

Commit:
Add encryption when caching Terminus sessions

- Update actions/cache to v4
- Add steps for encrypting/decrypting cached session
- Separate step just for caching Terminus plugins

Commit:
Fix missing cp, update README

Commit:
Move some steps around

Commit:
Fix cache directory path

Commit:
Update steps ids

Commit:
Update actions/checkout version for tests

Commit:
Use openssl vs gpg2 for platform compatibilities

- gpg2 is not installed on MacOS by default, openssl is available
- clean up some comments

Commit:
Add pbkdf2 and iter for improved strength

Address "deprecated key derivation used" output

Commit:
Fix encrypt step name reference

Commit:
Update authentication pattern

- Remove redundant auth step
- Use whoami to validate session
- Remove redundant file copying steps
- Remove Terminus plugins cache (can add back later if we support installing plugins)

Commit:
Remove restore key

kyletaylored commented:
I've integrated the source code from Lullabot's GitHub Action (#18) so that their action could be deprecated (talked with @penyaskito yesterday about this).

1. Cache implemented for sharing session across jobs
2. Session is encrypted using openssl / aes-256 before being stored into cache
3. Session is scoped to workflow run (re-running same workflow will use existing cache, new workflows create new session)

You can see a successful run of it here (see embedded steps in primary job and child jobs): https://github.com/the-scranton-branch/Scranton-WordPress/actions/runs/8206546810/job/22454187954

kporras07 reviewed (approved):
Looks great!

kyletaylored commented:
This also fixes #18 as it incorporates some of the code natively versus referencing it. Will close that PR.
